### ISSUE TITLE
 Exception when creating copy of `Option` object with `Constraint`

### ISSUE THREAD

**Issue by MePsyDuck** [janluke/cloup#166]
Using:
* python 3.10
* click 8.1.7
* cloup 3.0.0

#### Bug description
Any class that extends `cloup.constraints.Constraint` cannot be copied via `copy` package or `copy()` method.

#### To Reproduce
Here's a minimal example to reproduce the issue
```python
import copy
from cloup.constraints import Constraint

class MyConstraint(Constraint):
    def help(self, ctx):
        return 'help text'

    def check_values(self, params, ctx):
        pass

print(getattr(MyConstraint(), "__reduce_ex__", None)) # return non-None value
copy.copy(MyConstraint()) # produces error
MyConstraint().copy() # produces error
```

Output
```python
Traceback (most recent call last):
  File ".../test.py", line 12, in test
    copy.copy(MyConstraint())
  File ".../python3.10/copy.py", line 92, in copy
    rv = reductor(4)
TypeError: 'NoneType' object is not callable
```

A library that I use ([click-extra](https://github.com/kdeldycke/click-extra)) uses `deepmerge` to build a parameter tree. When attempting to make a deepcopy of the `Option` object, a copy of the `constraint` is necessary, but instead it results in the exception mentioned above.

Since my understanding of Python is limited, I'm unable to debug the issue on my own. However, I have managed to narrow down the problem to this point. If there's a way I can assist in resolving this issue, please let me know.

**Comment by janluke** [comment]
Thanks for reporting. I'm quite busy these days but I'll see what I can do this week. Otherwise, @kdeldycke feel free to have a look at it :)

**Comment by janluke** [comment]
Fix released in v3.0.1.

**Comment by kdeldycke** [comment]
Oh thank both of you for this! Thank you @MePsyDuck for taking the time to dig in into Click Extra internals, and post an issue here, even if you don't know Python. 🤗

I was not aware of this issue and used `deepmerge` a little bit blindly, not knowing its detailed implementation. But now, in light of @janluke fix I get it.

Huge respect for you @janluke for looking into it and cutting a release while being busy. 💪 Also busy myself so I'll make the effort to release a new bug-fix version of Click Extra too. 😅

**Comment by MePsyDuck** [comment]
Great to see the issue resolved in version 3.0.1! Thanks, @janluke , for the quick fix despite your busy schedule. 
Also, thank you, @kdeldycke , for your kind words! I'm glad I could contribute. Looking forward to continued improvements.

**Comment by janluke** [comment]
> But now, in light of @janluke fix I get it.

I actually don't have a full understanding of it :D. I replicated the issue and read an error message that didn't make any sense:
![image](https://github.com/janluke/cloup/assets/6934216/ffeef14b-4aaf-4e92-a376-646cd72017f2)

I just thought it might be related to `__getattr__` and noticed the mistake.
